### DISCUSSION TITLE
How to bind C++ resources into a Python context manager

### DISCUSSION THREAD

**Discussion by nicholasjng** [wjakob/nanobind#388] [Q&A] [answered]
Say I have a resource like a simple timer on a class, which makes sense to be acquired and released inside a context manager on the Python side:

```
  nb::class_<State>(m, "State")
      .def("pause_timing", &State::PauseTiming) # <- C++ state APIs
      .def("resume_timing", &State::ResumeTiming) #
```

How could I expose this on the Python side as a context manager, say like `with State().pause()`, where `State.pause().__enter__()` calls `State.pause_timing`, and `State.pause().__exit__()` calls `resume_timing`? Is there a convenience type for C++ resources to context managers in Python? 
If there is none and it's feasible to add, consider me signed up for contribution. Thanks!

**Comment by wjakob** [comment, marked as answer]
No abstraction for this exists, and there aren't plans to add one. This is how I would do it (likely there are some syntax errors, you will need to adapt it).
```cpp
struct PauseContext { State &state; };

auto state = nb::class_<State>(m, "State")
   .def(...);

nb::class_<PauseContext>(state, "PauseContext")
    .def("__enter__", [](PauseContext &ctx) { ctx.state.PauseTiming(); })
    .def("__exit__", [](PauseContext &ctx, nb::handle, nb::handle, nb::handle) { ctx.state.ResumeTiming(); },
         nb::arg().none(), nb::arg().none(), nb::arg().none());

state.def("pause", [](State &s) { return PauseContext{s}; }, nb::rv_policy::reference_internal);
```

**Reply by nicholasjng** [reply]
Thank you! That was easier than I expected.

**Reply by nicholasjng** [reply]
One more follow-up question if you allow: Is this way (i.e. binding a C++ object) preferred to a lightweight Python class wrapping around the already exposed `State` class? I read a bit about Python subclassing of C++ objects in the documentation, but I don't have a feeling for whether the ownership problems presented there would happen in this case.

**Reply by wjakob** [reply]
Both are fine, no issues with sharing are expected.
